### ISSUE TITLE
Looking for feedback: Adding support for multiple xrdp endpoint bindings?

### ISSUE THREAD

**Issue by jterry75** [neutrinolabs/xrdp#1124]
With the added support for `AF_VSOCK` some users would like to be able to connect both hypervisor local and also expose a remote endpoint at the same time. This would be a modification to the config to support both `tcp` settings and `vsock` settings for the same daemon at the same time. Then a user could connect via a fast connection (ip independent) when local but also connect to this vm when remote. This is roughly exactly what the windows `TermSrv` does to accomplish this. 

A few thoughts:
- We would need to modify the config to support connection independent settings and add new connection dependent settings sections. IE: you might want `bitmap_compression=true` for `tcp` but `=false` for `vsock`.
- The xrdp daemon would need to have a listener per connection entry and dispatch reads/writes with the appropriate settings.
- If a user disconnected from `tcp/vsock` and switched to the other they should see the same user session. (I don't think this is a problem just pointing it out).
- The same user account should be able to connect over both endpoints at the same time. Connecting via another endpoint should disconnect the previous connection and vice versa.

Other ideas/gotchas you can think of? I would be happy to try and work out some details in the code base here if everyone thinks its a good idea.

**Comment by worthwhileindustries** [comment]
What about not at the same time? I'm trying to use MSTSC to connect with one single connection but, I cannot figure it out. Is xrdp not using the same 3389 port?

**Comment by metalefty** [comment]
The main topic of this issue is vsock and tcp are exclusive, right? vsock is implemented by you and you make it exclusive. I'm OK if you make xrdp can listen both socket at the same time. Send a PR first.

> IE: you might want bitmap_compression=true for tcp but =false for vsock.

IMO, I don't like it. It will make config complex and config parser with little benefit.

**Comment by jterry75** [comment]
@worthwhileindustries - I'm not sure they are related to what I am asking. Check that your guest is configured in '/etc/xrdp/xrdp.ini' to use tcp and that your machine has that port open. Then you should be able to use mstsc to connect. We do this all the time.

@metalefty - Thanks for the feedback. I will work on getting a PR up supporting both connections but will not add the connection dependent settings for now. We can always do that later if it becomes an issue.

**Comment by jsorg71** [comment]
What is we get rid of xrdp.ini entry 'use_vsock' and just use the 'sock=' line in xrdp.ini

sock=/tmp/local.xrdp, tcp:192.168.1.1:3389, tcp:3390, tcp6:4500, vsock:1000

listen on unin domain socket /tmp/local.xrdp
listen on interface 192.168.1.1 tcp port 3389
listen on all interfaces tcp port 3390
listen on all interfaces tcp6 port 4500
listen on vsock port 1000

I think if you want different behavior for vsock vs tcp connection, there should be separate xrdp installs with a different prefix.  It would get too complicated otherwise.

**Comment by metalefty** [comment]
@jsorg71 The idea is good. The format needs to be considered.

Let's specify the socket type `uds` or `unix` for unix doman socket.

If I want to specify to listen on a specific IPv6 interface, it looks complicated :(

2001 is a port? or a part of IPv6 address?? 
```
sock= unix:/tmp/xrdp.socket tcp6:2001:foo:bar::buzz:3389
```

Still ugly :(
```
sock= tcp6:[2001:foo:bar::buzz]:3389
```

**Comment by jterry75** [comment]
The URI format for all these examples would be (I think):
```
unix://./tmp/xrdp.socket
tcp://.:3380/  (<- Localhost 127.0.0.1:3380)
tcp://:3380/ (<- *:3380)
tcp://<any ipv4 format addr>:3380) (<- 192.168.1.1:3380)
tcp6://.:3380/
tcp6://:3380/
tcp6://<any ipv6 format>:3380/
vsock://<cid>:<port>/
```

**Comment by jsorg71** [comment]
@jterry75 If you are going to do a PR, in xrdp_types.h, struct xrdp_listen, the 'listen_trans' needs to be changed to an array, or better yet, a list.  One struct_trans for each item in the 'sock=' line.

**Comment by jsorg71** [comment]
I created a working prototype using lists
https://github.com/jsorg71/xrdp/tree/listen-list

**Comment by jterry75** [comment]
@jsorg71 - This is awesome! I did a super quick CR but will do a more thorough one later. Will install this private and give it a shot but it looks like it will do exactly what I was thinking. Way cool

**Comment by jsorg71** [comment]
@jterry75 Did you get any time to work on this?  If not, I think I should clean this up and do a PR.

**Comment by jterry75** [comment]
Would you mind doing a rebase and PR. It did indeed work if I remember correctly

**Comment by beaukinstler** [comment]
Hello @jterry75 and @jsorg71. It looks like this was never rebased and merged, is that correct?

**Comment by jsorg71** [comment]
The items left to address.

1.  The old code was running a "listen test" that was used before the fork in daemon mode to make sure it can listen on the port.  Not sure what the new code should do,  Maybe test that all can listen? or at least one?
2.  Do we need to have code to still use the old xrdp.ini file format?  I have not done this yet.
3.  What about the -p parameter when starting xrdp.  -p is for port.  Maybe if -p is used, override xrdp.ini listen list and just listen on ::0:port ?

If we can figure these items out, we can merge.

**Comment by jsorg71** [comment]
I think all the items are addressed and I created PR #1366
